File: selected_coding_interview/codes/java/lc_287_find_the_duplicate_number/lc_287_find_the_duplicate_number_s2.java
/*
* File: lc_287_find_the_duplicate_number_s2.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_287_find_the_duplicate_number;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public int findDuplicate(int[] nums) {
            int i = 0;
            while(i < nums.length) {
                if(nums[i] == i) {
                    i++;
                    continue;
                }
                if(nums[nums[i]] == nums[i]) return nums[i];
                int tmp = nums[i];
                nums[i] = nums[tmp];
                nums[tmp] = tmp;
            }
            return -1;
        }
    }

public class lc_287_find_the_duplicate_number_s2 {
    public static void main(String[] args) {
        // ======= Test Case =======
        int[] test_input = new int[]{1, 3, 4, 2, 2};
        int expected_output = 2;

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.findDuplicate(test_input);
        System.out.println(result);

    }
}

File: selected_coding_interview/codes/python/lc_287_find_the_duplicate_number_s2.py
"""
File: lc_287_find_the_duplicate_number_s2.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        i = 0
        while i < len(nums):
            if nums[i] == i:
                i += 1
                continue
            if nums[nums[i]] == nums[i]: return nums[i]
            nums[nums[i]], nums[i] = nums[i], nums[nums[i]]
        return -1


# ======= Test Case =======
test_input = [1, 3, 4, 2, 2]
expected_output = 2

# ====== Driver Code ======
slt = Solution()
result = slt.findDuplicate(test_input)
print(result)
